contracts/utils/Checkpoints.sol

```diff
@@ -83,4 +68,314 @@ library Checkpoints {
     ) internal returns (uint256, uint256) {
         return push(self, op(latest(self), delta));
     }
+
+    function _insert(
+        Checkpoint[] storage self,
+        uint32 key,
+        uint224 value
+    ) private returns (uint224, uint224) {
+        uint256 pos = self.length;
+
+        if (pos > 0) {
+            // Use of memory is important here.
+            Checkpoint memory last = _unsafeAccess(self, pos - 1);
+
+            // Checkpoints keys must be increassing.
+            require(last._blockNumber <= key, "Checkpoint: invalid key");
+
+            // Update or push new checkpoint
+            if (last._blockNumber == key) {
+                _unsafeAccess(self, pos - 1)._value = value;
+            } else {
+                self.push(Checkpoint({_blockNumber: key, _value: value}));
+            }
+            return (last._value, value);
+        } else {
+            self.push(Checkpoint({_blockNumber: key, _value: value}));
+            return (0, value);
+        }
+    }
+
+    function _upperBinaryLookup(
```

Comment:
I think we should have at least minimal documentation for these functions, in this case for example we can say that the array is assumed to be sorted, and that high is assumed to be at most the length.